libsolidity/parsing/Parser.cpp

```diff
@@ -708,6 +727,59 @@ ASTPointer<UsingForDirective> Parser::parseUsingDirective()
 	return nodeFactory.createNode<UsingForDirective>(library, typeName);
 }
 
+ASTPointer<ModifierArea> Parser::parseModifierArea(ASTString const* _contractName, ASTPointer<ModifierArea> const& _parent)
+{
+	RecursionGuard recursionGuard(*this);
+	ASTNodeFactory nodeFactory(*this);
+
+	std::vector<ASTPointer<ModifierInvocation>> modifiers;
+
+	expectToken(Token::Using);
+	expectToken(Token::Modifier);
+
+	while (true)
+	{
+		modifiers.push_back(parseModifierInvocation());
+		if (m_scanner->currentToken() == Token::Comma)
+			m_scanner->next();
+		else
+		{
+			expectToken(Token::LBrace);
+			break;
+		}
+	};
+
+	ASTPointer<std::vector<ASTPointer<FunctionDefinition>>> functions = std::make_shared<std::vector<ASTPointer<FunctionDefinition>>>();
+	ASTPointer<std::vector<ASTPointer<ModifierArea>>> subAreas = std::make_shared<std::vector<ASTPointer<ModifierArea>>>();
+
+	ASTPointer<ModifierArea> modifierArea = nodeFactory.createNode<ModifierArea>(modifiers, functions, subAreas, _parent);
```

Comment:
OK - this is probably why you use ``ASTPointer``'s for the functions and subareas in ``ModifierArea`` - so that you can create the modifier area here and still fill in the functions and subareas below. However: are you sure you don't introduce cyclic dependencies of shared pointers here? The modifier areas contains a list of shared pointers to function definitions and the function definitions contain a shared pointer back to the modifier area? You probably at least need to change the parent pointers in the function definitions to a plain pointer instead of a shared pointer, although it would even be better if you could avoid such cyclic constructions altogether, if at all possible.